### PR TITLE
Struct output in verilog

### PR DESCRIPTION
**Pull request by darthscsi**
Output verilog for structures.
Also support merging declarations with common packed portion but different unpacked version.  eg:
```
input  [7:0]                                              uarray[15:0], postUArray,
```

### CODE REVIEW

File: lib/Translation/ExportVerilog/ExportVerilog.cpp [hunk 1]
```diff
@@ -166,6 +168,92 @@ static bool isZeroBitType(Type type) {
...
+        return stripUnpackedTypes(arrayType.getElementType());
+      })
+      .Default([](Type type) { return type; });
+}
+
+/// Output the basic type that consists of packed and primitive types.  This is
+/// those to the left of the name in verilog.
+static void printPackedTypeImpl(Type type, raw_ostream &os, Location loc,
+                                SmallVectorImpl<size_t> &dims,
+                                bool implicitIntType, char structFieldSep) {
```
[the remark this answers is not in the document]
Comment (teqdruid) [reply lib/Translation/ExportVerilog/ExportVerilog.cpp:186]: Maybe consider renaming `implicitIntType` to something along the lines of `keywordAlreadyPrinted`?

File: lib/Translation/ExportVerilog/ExportVerilog.cpp [hunk 2]
```diff
@@ -166,6 +168,92 @@ static bool isZeroBitType(Type type) {
...
+        dims.push_back(arrayType.getSize());
+        printPackedTypeImpl(arrayType.getElementType(), os, loc, dims,
+                            implicitIntType, structFieldSep);
+      })
+      .Case<InterfaceType>([](InterfaceType ifaceType) {
+        // Noop
+      })
+      .Case<UnpackedArrayType>([&](UnpackedArrayType arrayType) {
+        os << "<<unexpected unpacked array>>";
+        emitError(loc, "Unexpected unpacked array in packed type ")
```
Comment (teqdruid) [inline lib/Translation/ExportVerilog/ExportVerilog.cpp:226]: I think this can be an `assert` since one is supposed to call `stripUnpackedTypes` before calling this, yes?

**Review by teqdruid — APPROVED** [review]
This looks great! Please add an integration test.

File: lib/Translation/ExportVerilog/ExportVerilog.cpp (hunk 2 above)
Comment (darthscsi) [reply lib/Translation/ExportVerilog/ExportVerilog.cpp:226]: that won't handle packed<unpacked<whatever>>.  Strip will only handle unpacked<packed<whatever>>.

File: lib/Translation/ExportVerilog/ExportVerilog.cpp [hunk 3]
```diff
@@ -166,6 +168,92 @@ static bool isZeroBitType(Type type) {
...
+        }
+        os << '}';
+      })
+      .Case<ArrayType>([&](ArrayType arrayType) {
+        dims.push_back(arrayType.getSize());
+        printPackedTypeImpl(arrayType.getElementType(), os, loc, dims,
+                            implicitIntType, structFieldSep);
+      })
+      .Case<InterfaceType>([](InterfaceType ifaceType) {
+        // Noop
```
[the remark this answers is not in the document]
Comment (darthscsi) [reply lib/Translation/ExportVerilog/ExportVerilog.cpp:222]: The problem is that interfaces are routed through here, but also dealt as a verilog base name like wire and reg.  As long as wire and reg are ops not types, this is the least invasive way to handle this that I saw.

File: lib/Translation/ExportVerilog/ExportVerilog.cpp (hunk 2 above)
Comment (teqdruid) [reply lib/Translation/ExportVerilog/ExportVerilog.cpp:234]: Yep. I misread the strip implementation.

File: lib/Translation/ExportVerilog/ExportVerilog.cpp [hunk 4]
```diff
@@ -166,6 +168,99 @@ static bool isZeroBitType(Type type) {
...
+
+/// Output the basic type that consists of packed and primitive types.  This is
+/// those to the left of the name in verilog. implicitIntType controls whether
+/// to print a base type for (logic) for inteters or whether the caller will
+/// have handled this (with logic, wire, reg, etc).  structFieldSep is a
+/// character to be printed between struct fields.  Pretty printed structs will
+/// likely use newline, while inline structs will use spaces
+static void printPackedTypeImpl(Type type, raw_ostream &os, Location loc,
+                                SmallVectorImpl<size_t> &dims,
+                                bool implicitIntType, char structFieldSep) {
```
Comment (teqdruid) [inline lib/Translation/ExportVerilog/ExportVerilog.cpp:193]: Is the use case for `structFieldSep` formatting for a typedef? If so, I'm working on one (on top of this patch). With `structFieldSep = true`,

```
typedef struct packed {logic bar;
logic [3:0][7:0] arr;
} StructFoo;
```

Comment (teqdruid) [reply lib/Translation/ExportVerilog/ExportVerilog.cpp:193]: I'd encourage you to leave this feature for a future patch.

File: include/circt/Dialect/SV/SVInOutOps.td [hunk 5]
```diff
@@ -48,6 +48,12 @@ def WireOp : SVOp<"wire", [DeclareOpInterfaceMethods<OpAsmOpInterface>]> {
 
   let assemblyFormat = "custom<ImplicitSSAName>(attr-dict) `:` type($result)";
   let hasCanonicalizer = true;
+
+  let extraClassDeclaration = [{
+    Type getElementType() {
+      return result().getType().cast<InOutType>().getElementType();
```
Comment (lattner) [inline include/circt/Dialect/SV/SVInOutOps.td:54]: I think that getType() should be typed as returning an inout type correctly, so this can just be:

`return getType().getElementType()`

Also the one for RegOp

File: lib/Translation/ExportVerilog/ExportVerilog.cpp [hunk 6]
```diff
@@ -86,6 +86,8 @@ static void getTypeDims(SmallVectorImpl<int64_t> &dims, Type type,
     return getTypeDims(dims, uarray.getElementType(), loc);
   if (type.isa<InterfaceType>())
     return;
+  if (type.isa<StructType>())
```
Comment (lattner) [inline lib/Translation/ExportVerilog/ExportVerilog.cpp:89]: nit: you can merge this into the previous if with `||`

File: lib/Translation/ExportVerilog/ExportVerilog.cpp [hunk 7]
```diff
@@ -166,6 +168,99 @@ static bool isZeroBitType(Type type) {
...
+      .Case<UnpackedArrayType>([](UnpackedArrayType arrayType) {
+        return stripUnpackedTypes(arrayType.getElementType());
+      })
+      .Default([](Type type) { return type; });
+}
+
+/// Output the basic type that consists of packed and primitive types.  This is
+/// those to the left of the name in verilog. implicitIntType controls whether
+/// to print a base type for (logic) for inteters or whether the caller will
+/// have handled this (with logic, wire, reg, etc).  structFieldSep is a
```
Comment (lattner) [inline lib/Translation/ExportVerilog/ExportVerilog.cpp:188]: `structFieldSep` looks like premature generalization to me, since it is always ' '.  i'd drop it and constant fold that through the code to simplify it.  The pretty printed types are likely to need more than just a newline (they'll want indentation too) so I don't think a char is going to be enough to handle this.

File: lib/Translation/ExportVerilog/ExportVerilog.cpp [hunk 8]
```diff
@@ -166,6 +168,99 @@ static bool isZeroBitType(Type type) {
...
+            os << "/*Zero Width*/";
+      })
+      .Case<InOutType>([&](InOutType inoutType) {
+        printPackedTypeImpl(inoutType.getElementType(), os, loc, dims,
+                            implicitIntType, structFieldSep);
+      })
+      .Case<StructType>([&](StructType structType) {
+        os << "struct packed {";
+        for (auto &element : structType.getElements()) {
+          SmallVector<size_t, 8> structDims;
```
Comment (lattner) [inline lib/Translation/ExportVerilog/ExportVerilog.cpp:216]: needless microoptimization: I'd recommend pull the smallvector out of the loop and clear it on each iteration instead of reinitializing it.

File: lib/Translation/ExportVerilog/ExportVerilog.cpp (hunk 2 above)
Comment (lattner) [reply lib/Translation/ExportVerilog/ExportVerilog.cpp:234]: Yeah this also catches the case if unpacked things end up in structs.

File: lib/Translation/ExportVerilog/ExportVerilog.cpp [hunk 9]
```diff
@@ -1115,6 +1213,42 @@ SubExprInfo ExprEmitter::visitComb(MuxOp op) {
...
+  StructType stype = op.getType().cast<StructType>();
+  os << "'{";
+  size_t i = 0;
+  llvm::interleaveComma(stype.getElements(), os,
+                        [&](const StructType::FieldInfo &field) {
+                          os << field.name << ": ";
+                          emitSubExpr(op.getOperand(i++), Selection);
+                        });
+  os << '}';
+  return {Unary, IsUnsigned};
```
Comment (lattner) [inline lib/Translation/ExportVerilog/ExportVerilog.cpp:1226]: It would be fine to split this off to a separate patch, but I think it would make sense to add an `IsStruct` case to `SubExprSignResult`.

File: lib/Translation/ExportVerilog/ExportVerilog.cpp [hunk 10]
```diff
@@ -1115,6 +1213,42 @@ SubExprInfo ExprEmitter::visitComb(MuxOp op) {
   return {Conditional, signedness};
 }
 
+SubExprInfo ExprEmitter::visitTypeOp(StructCreateOp op) {
+  StructType stype = op.getType().cast<StructType>();
```
Comment (lattner) [inline lib/Translation/ExportVerilog/ExportVerilog.cpp:1217]: This `cast` shouldn't be needed.

File: lib/Translation/ExportVerilog/ExportVerilog.cpp [hunk 11]
```diff
@@ -1115,6 +1213,42 @@ SubExprInfo ExprEmitter::visitComb(MuxOp op) {
...
+SubExprInfo ExprEmitter::visitTypeOp(StructInjectOp op) {
+  StructType stype = op.getType().cast<StructType>();
+  os << "'{";
+  llvm::interleaveComma(stype.getElements(), os,
+                        [&](const StructType::FieldInfo &field) {
+                          os << field.name << ": ";
+                          if (field.name == op.field()) {
+                            emitSubExpr(op.newValue(), Selection);
+                          } else {
+                            emitSubExpr(op.input(), Selection);
```
Comment (lattner) [inline lib/Translation/ExportVerilog/ExportVerilog.cpp:1244]: This is emitting `op.input()` mutliple times which can lead to expontential blowup in generated verilog.  Please make sure this is handled in isExpressionUnableToInline like SExtOp

**Review by lattner — APPROVED** [review]
This is really really nice, great work

**Comment by teqdruid** [comment]
This doesn't seem to cover expressions which aren't forward declared.

```
rtl.module @M1(%val : i8) -> (%out: !rtl.struct<b: i8>)  {
  %byte = rtl.struct_create (%val) : !rtl.struct<b: i8>
  sv.fwrite "%d\n"(%byte) : !rtl.struct<b: i8>
  rtl.output %byte : !rtl.struct<b: i8>
}
```
produces
```
...
wire _T = '{b: val};
...
```

I was just experimenting with the typedefop I'm building on top of this (using the new type printing stuff), and it turns out that when you typedef a logic and use it at the module level, it doesn't result in a continuous assignment as a wire does. You need to prefix it with `wire` to make it a continuous assignment. This is valid and what we want:

```
typedef logic [7:0] Byte;

module B(
  input Byte b);

  wire Byte dst = b;
endmodule
```

That may have been obvious to you. I don't know if the same is true for structs.

**Comment by darthscsi** [comment]
> This doesn't seem to cover expressions which aren't forward declared.
> 
> ```
> ...
> wire _T = '{b: val};
> ...
> ```

I'll see about forcing these to named and typed temporaries.

> I was just experimenting with the typedefop I'm building on top of this (using your new type printing stuff), and it turns out that when you typedef a logic and use it at the module level, it doesn't result in a continuous assignment as a wire does. You need to prefix it with `wire` to make it a continuous assignment. This is valid and what we want:

Verilog makes me sad sometimes.

**Comment by teqdruid** [comment]
Sometimes?

File: include/circt/Dialect/SV/SVInOutOps.td (hunk 5 above)
Comment (darthscsi) [reply include/circt/Dialect/SV/SVInOutOps.td:54]: It is not currently.  Looks like it is just a constraint on the type and doesn't result in the storage type changing.

**Comment by lattner** [comment]
nice!!
